### ISSUE TITLE
[AArch64] Renaissance benchmark hangs when Graal is used

### ISSUE THREAD

**Issue by pli-arm** [oracle/graal#2045]
Recently, we noticed that on some types of AArch64 CPUs, the Renaissance benchmark hangs when Graal is used as the JVMCI compiler. This issue can be reproduced by below test command on Graal master with jdk11u.
```
cd graal/compiler
mx vm -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -XX:+UseJVMCICompiler \
  -Djvmci.Compiler=graal -jar renaissance-gpl-0.10.0.jar future-genetic
```
where renaissance-gpl-0.10.0.jar can be downloaded by
```
wget https://github.com/renaissance-benchmarks/renaissance/releases/download/v0.10.0/renaissance-gpl-0.10.0.jar
```

We have some investigations on this issue. Below are some details. TL;DR

The future-genetic case in the Renaissance benchmark is a test which makes heavy use of the `java.util.concurrent` library. It submits a map-reduce job but hangs with a phenomenon that all the worker threads are parked. The hanging issue only appears on AArch64 when we use Graal as the JVMCI compiler. If we just use C1 or C2, everything is fine. This issue is very hard to be reproduced in a small Java program. As we noticed that the library code called by this case has a lot of volatile accesses, we guessed that Graal might have something wrong when compiling loads/stores with volatile.

To prove our guess, we dumpped the assembly code compiled for volatile load by each compiler. The C1 compiler uses a normal load instruction surrounded by two dmb instructions for a volatile load. Code is like
```
dmb ish
ldr [xxx]
dmb ishld
```
while C2 has optimized the volatile load to a below AArch64 load-acquire instruction.
```
ldar [xxx]
```
In the Graal compiler, we only generates one dmb after the load, like below. There's no dmb before.
```
ldr [xxx]
dmb ishld
```

To test that the missing DMB is the cause of hanging, we tried to modify the Graal's compiled to what C1's and/or C2's like. We found that, either by hard-coding one more DMB before the Graal compiled LDR or by replacing the LDR+DMB to a single LDAR, the hanging issue disappers. To see why C1 needs one more DMB before the load, we found JDK-8179954 (https://bugs.openjdk.java.net/browse/JDK-8179954) stating that the mixed use of STLR and LDR+DMB may cause some errors. We also found this problem mentioned in the comments of @adinn's PR #1772.

To confirm this is a similar issue with JDK-8179954, we searched the code-gen and found there's no STLR generated by current Graal. But we do have another similar instruction, STLXR, generated for atomic operations. Suppose that an STLXR is generated before any LDR+DMB, the STLXR and LDR may be reordered by CPU according to the AArch64 memory ordering semantics. But if the STLXR represents an atomic store of a volatile field and the following LDR+DMB represents another volatile load, the reordering of these is prohibited according to the JSR-133 Cookbook (http://gee.cs.oswego.edu/dl/jmm/cookbook.html). We also tried adding one DMB after each STLXR generated for atomics, this hanging issue also disappers without any surprise.

Above are all our investigations. We believe that this issue will disapper after #1772 gets merged. So we would like to ask is there anything we (Arm) could do to make #1772 reviewed and merged ASAP? If it is not possible to be merged in a short time, shall we add a temporary workaround patch (just adding one more DMB somewhere) to make the `java.util.concurrent` library work correctly?

**Comment by adinn** [comment]
Hi Pengfei,

> To confirm this is a similar issue with JDK-8179954, we searched the code-gen and found there's no STLR generated by current Graal. But we do have another similar instruction, STLXR, generated for atomic operations. Suppose that an STLXR is generated before any LDR+DMB, the STLXR and LDR may be reordered by CPU according to the AArch64 memory ordering semantics. But if the STLXR represents an atomic store of a volatile field and the following LDR+DMB represents another volatile load, the reordering of these is prohibited according to the JSR-133 Cookbook (http://gee.cs.oswego.edu/dl/jmm/cookbook.html). We also tried adding one DMB after each STLXR generated for atomics, this hanging issue also disappers without any surprise.

Your analysis of this case is correct. It is not sufficient to use a plain stlxr to implement atomic writes to a volatile Java field since that does not ensure that subsequent volatile loads are ordered relative to the atomic write. Given that there is currently no use of stlr/ldar it would be to mots appropriate to implement the atomic operations with a stxr combined with a preceding and trailing dmb ish i.e. to implement it using the standard barrier-based pattern.  However, using stlxr with a trailing dmb ish would also achieve the same result.

However, I am not sure whether it will be correct to remove the barriers and revert the atomic operations to simply use stlxr once #1772 is merged. That would be ok to implement atomic updates to Java fields. However, there may be a problem if that translation is used for atomic updates to non-Java field values. Can you be more specific about which atomic updates you are talking about?

**Comment by XiaohongGong** [comment]
Hi adinn，

> However, there may be a problem if that translation is used for atomic updates to non-Java field values.

Could you give an example of the atomic for "non-java" field? And for the non-java field atomic, do you mean it still has issues when we use "stlxr" for atomic and "ldar" for volatile load? Thank you!

Xiaohong

**Comment by XiaohongGong** [comment]
Hi adinn,

> Can you be more specific about which atomic updates you are talking about?

It's the `CompareAndSwapOp` (CAS) that can make this issue happen.

Thanks,
Xiaohong

**Comment by adinn** [comment]
> Could you give an example of the atomic for "non-java" field? And for the non-java field atomic, do you mean it still has issues when we use "stlxr" for atomic and "ldar" for volatile load? Thank you!

No, I cannot give an example of the atomic for non-java field. I was asking you if you could be sure that no such example exists. The reason for asking that is that the corrrectness of the ldar/stlr translation is very closely tied to the semantics of the Java memory model. The model requires volatile accesses to order visibility of  intervening Java memory accesses across threads. The ldar/stlr translation provides that guarantee. However, what it fails to do is guarantee that the stlr write is ordered before later writes or reads to non-volatile field locations.

That may not be a strong enough guarantee for all the code which relies on atomic operations. In OpenJDK atomic operations are used for both Java program atomic operations and also internal to the VM, The latter sometimes need the implementation to provide a full barrier (see [this thread](https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2019-November/008164.html) where I had assumed generated atomic operations were only used for Java program operations and was corrected by Andrew Haley with an example showing showing where the full semantics are needed), That is why I was asking.

> It's the CompareAndSwapOp (CAS) that can make this issue happen.

Well, currently CompareAndSwapOp nodes are created in the back end by class DefaultJavaLoweringProvider under method lowerCompareAndExchangeNode() and by class WordOperationPlugin under method casOp(). I think the lowering only occurs for nodes created in response to Java application program operations (as opposed to Java code implementing e.g. SubstrateVM operations) so Java semantics may be adequate but that really needs checking very carefully.

I am more concerned that WordOpeation.casOp() is executed in snippet code (these are similar to intrinsics) created to implement Unsafe.CAS operations. Those calls may be used to implement various 'magic' VM operations. So, even though these snippets exist to replace what is /nominally/ Java code the caller may still be relying on the full barrier semantics. That might not be true now but it might change in the future. For example, current or future GC operations implemented using snippets may need to guarantee that an Unsafe.CAS of a pointer into an object header is visible before writing values into object fields.

**Comment by XiaohongGong** [comment]
Hi adinn,

I understand what you worried about after looking at [thread](https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2019-November/008164.html) .  Thank you! I will check these atomic operations in Graal and have a patch to fix them like hotspot if needed.

Thanks,
Xiaohong

**Comment by XiaohongGong** [comment]
Hi adinn,

As you described above, WordOperation.casOp() is used inside the snippet code and SubstrateVM. Here is its three usages:
1) MonitorSnippet both for hotspot and SubstrateVM
    This is mainly used for checking and getting the mark part of the lock object. I compared it with C2, and found that C2 also uses "ldaxr, stlxr" for monitor without any full barrier.
2) Addition for a JNI JavaVM in SubstrateVM.
   Code is in [JNIJavaVMList.java](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.jni/src/com/oracle/svm/jni/JNIJavaVMList.java)
3) Isolate creation and the heap initialization in SubstrateVM.
  Code is in [CEntryPointSnippets.java](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core.graal/src/com/oracle/svm/core/graal/snippets/CEntryPointSnippets.java) and [LinuxImageHeapProvider.java](https://github.com/oracle/graal/blob/master/substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/linux/LinuxImageHeapProvider.java)

I'm not sure whether we must use the conservative memory barrier for these usages. But I think maybe the caller can make the decision. So my question is: do you think if it's better to add a flag for the caller to choose whether to use the conservative barrier for the atomic or not.  The compiler can implement the atomic according to the flag.  And by default, it is the non-conservative pattern (the status now). 

Thanks,
Xiaohong

**Comment by adinn** [comment]
Hi Xiaohong,

I am not 100% sure what these 3 cases are doing but I think it is necessary for the default behaviour to remain conservative. That is not only to allow for what these cases might be doing but also to avoid problems with any future cases. If you want to optimize compare and swap operations for Java volatile fields then I suggest you modify the handling of Java fields to introduce a different form of CAS node which can be handled i the back end by omitting the barrier and relying on just the ldaxr/stlxr pair. You may be able to do that without introducing a new node by basing the decision on the location type of the CAS.

@theRealAph may want to comment on this.

**Comment by XiaohongGong** [comment]
Hi @adinn ,

I have a PR to fix this issue here: https://github.com/oracle/graal/pull/2181. The main idea of it is:
1) For non-java (VM internal) atomic operations, we always use the conservative barriers.
2) For java field atomic, we use "UseBarriersForVolatile" option to decide whether to use conservative barriers or not. This option is also added for SubstrateVM.
3) Temporarily insert a full barrier after STLXR of the normal atomics to make the Renaissance benchmark work well. This can be removed once LDAR is used for java volatile read.

So could you please take a look at it? If there are any comments please let me know. Thank you so much!

Thanks,
Xiaohong

**Comment by pli-arm** [comment]
I'm closing this as PR-1772 is merged.
